### PR TITLE
Rest client generator

### PR DESCRIPTION
**Pull request by purple52**

### CODE REVIEW

File: messaging/messaging-core/src/test/java/uk/gov/justice/services/messaging/JsonObjectEnvelopeConverterTest.java [hunk 1]
```diff
@@ -71,31 +74,78 @@ public void shouldReturnEnvelope() throws Exception {
...
+
+        JsonArrayBuilder jsonArrayBuilder = Json.createArrayBuilder();
+        jsonArrayBuilder.add(ARRAY_ITEM_1);
+        jsonArrayBuilder.add(ARRAY_ITEM_2);
+
+        return DefaultJsonEnvelope.envelopeFrom(JsonObjectMetadata.metadataFrom(metadataBuilder.build()), jsonArrayBuilder.build());
+    }
 
+    private JsonEnvelope envelopeWithNumberPayload() {
         JsonObjectBuilder metadataBuilder = Json.createObjectBuilder();
```
Comment (matt-rich) [inline messaging/messaging-core/src/test/java/uk/gov/justice/services/messaging/JsonObjectEnvelopeConverterTest.java]: metabuilder section is common, extract to private method?

File: messaging/messaging-core/src/main/java/uk/gov/justice/services/messaging/JsonObjectEnvelopeConverter.java [hunk 2]
```diff
@@ -4,13 +4,19 @@
...
+
+import static javax.json.JsonValue.ValueType.ARRAY;
+import static javax.json.JsonValue.ValueType.OBJECT;
 
 /**
  * A converter class to convert between {@link JsonEnvelope} and {@link JsonObject}.
  */
 public class JsonObjectEnvelopeConverter {
 
     public static final String METADATA = "_metadata";
```
Comment (matt-rich) [inline messaging/messaging-core/src/main/java/uk/gov/justice/services/messaging/JsonObjectEnvelopeConverter.java]: Does this need to be public? and if yes is it worthwhile having RESULTS public also?

File: core/src/test/java/uk/gov/justice/services/core/annotation/ComponentTest.java [hunk 3]
```diff
@@ -82,4 +84,29 @@ public void shouldThrowExceptionIfTierInvalid() {
...
+        assertTrue(Component.contains("COMMAND_HANDLER"));
+        assertTrue(Component.contains("EVENT_LISTENER"));
+        assertTrue(Component.contains("EVENT_PROCESSOR"));
+        assertTrue(Component.contains("QUERY_API"));
+        assertTrue(Component.contains("QUERY_CONTROLLER"));
+        assertTrue(Component.contains("QUERY_VIEW"));
+    }
+
+    @Test
+    public void shouldReturnTrueIfDoesNotContainGivenString() {
```
Comment (matt-rich) [inline core/src/test/java/uk/gov/justice/services/core/annotation/ComponentTest.java]: Should this test name be shouldReturnFALSE ?

File: core/src/main/java/uk/gov/justice/services/core/annotation/Component.java [hunk 4]
```diff
@@ -73,6 +76,19 @@ public static Component componentFrom(final InjectionPoint injectionPoint) {
...
+    /**
+     * Checks if enum contains component with given name
+     * @param name - component name
+     * @return true if the enum contains the given name, false otherwise
+     */
+    public static boolean contains(final String name) {
+        return valuesStream().filter(c -> c.name().equals(name)).findAny().isPresent();
+    }
+
+    private static Stream<Component> valuesStream() {
```
Comment (matt-rich) [inline core/src/main/java/uk/gov/justice/services/core/annotation/Component.java]: Does this (and other utility methods) need to be public?

File: components/query/query-api/pom.xml [hunk 5]
```diff
@@ -18,8 +18,13 @@
             <version>${project.version}</version>
         </dependency>
         <dependency>
+        <groupId>uk.gov.justice.services</groupId>
```
Comment (matt-rich) [inline components/query/query-api/pom.xml]: indentation is incorrect

File: common/src/main/java/uk/gov/justice/services/common/converter/JsonValueToStringConverter.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+public class JsonValueToStringConverter implements Converter<JsonValue, String> {
+
+    @Override
+    public String convert(final JsonValue source) {
+        final StringWriter stringWriter = new StringWriter();
+        try (final JsonWriter writer = Json.createWriter(stringWriter)) {
+
+            final JsonValue.ValueType valueType = source.getValueType();
+
+            if (valueType == JsonValue.ValueType.OBJECT || valueType == JsonValue.ValueType.ARRAY) {
```
Comment (matt-rich) [inline common/src/main/java/uk/gov/justice/services/common/converter/JsonValueToStringConverter.java]: static imports

File: common/src/main/java/uk/gov/justice/services/common/converter/JsonValueToStringConverter.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+/**
+ * Converts JsonValue to Json as a String.
+ */
+public class JsonValueToStringConverter implements Converter<JsonValue, String> {
+
+    @Override
+    public String convert(final JsonValue source) {
+        final StringWriter stringWriter = new StringWriter();
+        try (final JsonWriter writer = Json.createWriter(stringWriter)) {
```
Comment (matt-rich) [inline common/src/main/java/uk/gov/justice/services/common/converter/JsonValueToStringConverter.java]: Can you create StringWriter in the try block also for auto closure?

File: clients/rest-client-generator/src/test/java/uk/gov/justice/services/clients/rest/generator/RestClientGenerator_MethodBodyTest.java [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+    public void before() {
+
+        restClientGenerator = new RestClientGenerator();
+        compiler = new JavaCompilerUtil(outputFolder.getRoot(), outputFolder.getRoot());
+    }
+
+    @Test
+    public void shouldCallRestClientWithEndpointDefinitionContainingBaseUri() throws Exception {
+
+        restClientGenerator.run(
```
Comment (matt-rich) [inline clients/rest-client-generator/src/test/java/uk/gov/justice/services/clients/rest/generator/RestClientGenerator_MethodBodyTest.java]: Is it possible to move some of this common code into a private method?

File: messaging/messaging-core/src/test/java/uk/gov/justice/services/messaging/JsonObjectEnvelopeConverterTest.java (hunk 1 above)
Comment (ZeeshanGhalib) [inline messaging/messaging-core/src/test/java/uk/gov/justice/services/messaging/JsonObjectEnvelopeConverterTest.java]: Done

File: clients/rest-client-generator/src/test/java/uk/gov/justice/services/clients/rest/generator/RestClientGenerator_CodeStructureTest.java [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+        assertThat(applicationClass.getAnnotation(ServiceComponent.class), not(nullValue()));
+        assertThat(applicationClass.getAnnotation(ServiceComponent.class).value().toString(), is("QUERY_API"));
+        assertBaseUriField(applicationClass.getDeclaredField("BASE_URI"));
+        assertRestClientField(applicationClass.getDeclaredField("restClientProcessor"));
+        assertRestClientHelperField(applicationClass.getDeclaredField("restClientHelper"));
+    }
+
+    @Test
+    public void shouldGenerateClassWithQueryControllerAnnotation() throws Exception {
+        restClientGenerator.run(
```
Comment (matt-rich) [inline clients/rest-client-generator/src/test/java/uk/gov/justice/services/clients/rest/generator/RestClientGenerator_CodeStructureTest.java]: repeated code again, can this be a private method.. or even a base class if this is common across tests.

File: messaging/messaging-core/src/main/java/uk/gov/justice/services/messaging/JsonObjectEnvelopeConverter.java (hunk 2 above)
Comment (ZeeshanGhalib) [inline messaging/messaging-core/src/main/java/uk/gov/justice/services/messaging/JsonObjectEnvelopeConverter.java]: I have made both of them private.

File: clients/rest-client-generator/src/main/java/uk/gov/justice/services/clients/rest/generator/RestClientGenerator.java [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+import static com.squareup.javapoet.TypeSpec.classBuilder;
+import static java.lang.String.format;
+import static javax.lang.model.element.Modifier.FINAL;
+import static javax.lang.model.element.Modifier.PUBLIC;
+import static org.apache.commons.lang3.StringUtils.capitalize;
+import static org.apache.commons.lang3.StringUtils.isEmpty;
+
+
+/**
+ * Generates code for the rest client. Some things to keep in mind for understanding
```
Comment (matt-rich) [inline clients/rest-client-generator/src/main/java/uk/gov/justice/services/clients/rest/generator/RestClientGenerator.java]: Might want to tidy up this head :)

File: clients/rest-client-generator/src/main/java/uk/gov/justice/services/clients/rest/generator/RestClientGenerator.java [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+public class RestClientGenerator implements Generator {
+
+
+    private static final String SERVICE_COMPONENT_PROPERTY = "serviceComponent";
+    private static final String REST_CLIENT_HELPER = "restClientHelper";
+    private static final String REST_CLIENT_PROCESSOR = "restClientProcessor";
+
+    @Override
+    public void run(final Raml raml, final GeneratorConfig generatorConfig) {
+        TypeSpec.Builder classSpec = classSpecOf(raml, generatorConfig);
```
Comment (matt-rich) [inline clients/rest-client-generator/src/main/java/uk/gov/justice/services/clients/rest/generator/RestClientGenerator.java]: static import

File: core/src/main/java/uk/gov/justice/services/core/annotation/Component.java (hunk 4 above)
Comment (ZeeshanGhalib) [inline core/src/main/java/uk/gov/justice/services/core/annotation/Component.java]: valuesStream() is a private method.

File: components/query/query-api/pom.xml (hunk 5 above)
Comment (ZeeshanGhalib) [inline components/query/query-api/pom.xml]: Fixed.

File: clients/rest-client-generator/src/main/java/uk/gov/justice/services/clients/rest/generator/RestClientGenerator.java [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+    }
+
+    private void generateCodeForResource(Resource resource, TypeSpec.Builder classBuilder) {
+        resource.getActions().forEach((actionType, action) -> generateCodeForAction(resource, action, classBuilder));
+    }
+
+    private void generateCodeForAction(Resource resource, Action action, TypeSpec.Builder classBuilder) {
+        switch (action.getType()) {
+            case GET:
+                Response response = action.getResponses().get("200");
```
Comment (matt-rich) [inline clients/rest-client-generator/src/main/java/uk/gov/justice/services/clients/rest/generator/RestClientGenerator.java]: Use HttpStatus.SUCCESS not 200

File: common/src/main/java/uk/gov/justice/services/common/converter/JsonValueToStringConverter.java (hunk 6 above)
Comment (ZeeshanGhalib) [inline common/src/main/java/uk/gov/justice/services/common/converter/JsonValueToStringConverter.java]: Done.

File: clients/rest-client-generator/src/main/java/uk/gov/justice/services/clients/rest/generator/RestClientGenerator.java [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+    }
+
+    private FieldSpec baseUriStaticFieldOf(final Raml raml) {
+        return FieldSpec.builder(String.class, "BASE_URI")
+                .addModifiers(Modifier.PRIVATE, FINAL, Modifier.STATIC)
+                .initializer("$S", raml.getBaseUri())
+                .build();
+    }
+
+    private void generateCodeForResource(Resource resource, TypeSpec.Builder classBuilder) {
```
Comment (matt-rich) [inline clients/rest-client-generator/src/main/java/uk/gov/justice/services/clients/rest/generator/RestClientGenerator.java]: final method params on signature

File: clients/rest-client-generator/src/main/java/uk/gov/justice/services/clients/rest/generator/RestClientGenerator.java [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+            case POST:
+                action.getBody().values().iterator().forEachRemaining(
+                        mimeType -> classBuilder.addMethod(methodOf(resource, action, mimeType)));
+                break;
+            default:
+                throw new IllegalStateException(format("Unsupported action type %s", action.getType()));
+        }
+    }
+
+    private MethodSpec methodOf(Resource resource, Action action, MimeType mimeType) {
```
Comment (matt-rich) [inline clients/rest-client-generator/src/main/java/uk/gov/justice/services/clients/rest/generator/RestClientGenerator.java]: final method params on signature

File: clients/rest-client-generator/src/main/java/uk/gov/justice/services/clients/rest/generator/RestClientGenerator.java [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+            builder.returns(methodReturnType);
+            builder.addStatement("return $L.request(def, envelope)", REST_CLIENT_PROCESSOR);
+        } else {
+            builder.addStatement("$L.request(def, envelope)", REST_CLIENT_PROCESSOR);
+        }
+
+        return builder.build();
+    }
+
+    private void addQueryParam(MethodSpec.Builder builder, QueryParameter parameter, String name) {
```
Comment (matt-rich) [inline clients/rest-client-generator/src/main/java/uk/gov/justice/services/clients/rest/generator/RestClientGenerator.java]: final method params on signature - also check other methods below

File: clients/rest-client-generator/src/main/java/uk/gov/justice/services/clients/rest/generator/RestClientGenerator.java [hunk 16]
```diff
@@ -0,0 +1,206 @@
...
+        String baseName = WordUtils.capitalize(header.replaceAll("[\\W_]", " ")).replaceAll("[\\W_]", "");
+
+        String actionTypeStr = actionType.name().toLowerCase();
+        baseName = baseName.substring(0, 1).toUpperCase() + baseName.substring(1);
+        return actionTypeStr + baseName;
+    }
+
+    private String classNameOf(String baseUri) {
+        String[] pathSegments = baseUri.split("/");
+        if (pathSegments.length != 8) {
```
Comment (matt-rich) [inline clients/rest-client-generator/src/main/java/uk/gov/justice/services/clients/rest/generator/RestClientGenerator.java]: Can we make 8 a constant please

File: clients/rest-client-generator/src/main/java/uk/gov/justice/services/clients/rest/generator/RestClientGenerator.java [hunk 17]
```diff
@@ -0,0 +1,206 @@
...
+            throw new IllegalArgumentException("baseUri must have 8 parts");
+        }
+        return format("Remote%s%s%s",
+                capitalize(pathSegments[7]),
```
Comment (matt-rich) [inline clients/rest-client-generator/src/main/java/uk/gov/justice/services/clients/rest/generator/RestClientGenerator.java]: Move magic numbers to meaningfully named constants.

File: clients/rest-client-generator/pom.xml [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+        <dependency>
+            <groupId>uk.gov.justice.services</groupId>
+            <artifactId>rest-client-core</artifactId>
+            <version>${project.version}</version>
+        </dependency>
+
+        <dependency>
+            <groupId>com.squareup</groupId>
+            <artifactId>javapoet</artifactId>
+            <version>1.6.1</version>
```
Comment (matt-rich) [inline clients/rest-client-generator/pom.xml]: Version should be moved up into parent pom.

File: clients/rest-client-core/src/main/java/uk/gov/justice/services/clients/core/RestClientProcessor.java [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+
+import javax.inject.Inject;
+import javax.json.JsonObject;
+import javax.ws.rs.client.Client;
+import javax.ws.rs.client.ClientBuilder;
+import javax.ws.rs.client.Invocation;
+import javax.ws.rs.client.WebTarget;
+import javax.ws.rs.core.Response;
+
+public class RestClientProcessor {
```
Comment (matt-rich) [inline clients/rest-client-core/src/main/java/uk/gov/justice/services/clients/core/RestClientProcessor.java]: Class comment missing

File: clients/rest-client-core/src/main/java/uk/gov/justice/services/clients/core/RestClientHelper.java [hunk 20]
```diff
@@ -0,0 +1,28 @@
+package uk.gov.justice.services.clients.core;
+
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import static java.util.Arrays.stream;
+
+public class RestClientHelper {
```
Comment (matt-rich) [inline clients/rest-client-core/src/main/java/uk/gov/justice/services/clients/core/RestClientHelper.java]: Class comment missing

File: clients/rest-client-core/src/main/java/uk/gov/justice/services/clients/core/RestClientHelper.java [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+    private static final String PATH_PARAM_PATTERN = "\\{(\\w+)\\}";
+
+    public Set<String> extractPathParametersFromPath(String path) {
```
Comment (matt-rich) [inline clients/rest-client-core/src/main/java/uk/gov/justice/services/clients/core/RestClientHelper.java]: final method signature param

File: clients/rest-client-core/src/main/java/uk/gov/justice/services/clients/core/EndpointDefinition.java [hunk 22]
```diff
@@ -0,0 +1,35 @@
...
+public class EndpointDefinition {
+
+    private String baseURi;
+    private String path;
+    private Set<String> pathParams;
+    private Set<QueryParam> queryParams;
+
+    public EndpointDefinition(String baseURi, String path, Set<String> pathParams,
+                              Set<QueryParam> queryParams) {
+        this.baseURi = baseURi;
```
Comment (mapingo) [inline clients/rest-client-core/src/main/java/uk/gov/justice/services/clients/core/EndpointDefinition.java]: arguments should be final

File: common/src/main/java/uk/gov/justice/services/common/converter/JsonValueToStringConverter.java (hunk 7 above)
Comment (ZeeshanGhalib) [inline common/src/main/java/uk/gov/justice/services/common/converter/JsonValueToStringConverter.java]: Done

File: clients/rest-client-core/src/main/java/uk/gov/justice/services/clients/core/QueryParam.java [hunk 23]
```diff
@@ -0,0 +1,20 @@
+package uk.gov.justice.services.clients.core;
+
+public class QueryParam {
+
+    private String name;
+    private boolean required;
+
+    public QueryParam(String name, boolean required) {
+        this.name = name;
```
Comment (mapingo) [inline clients/rest-client-core/src/main/java/uk/gov/justice/services/clients/core/QueryParam.java]: final

File: clients/rest-client-core/src/main/java/uk/gov/justice/services/clients/core/RestClientHelper.java [hunk 24]
```diff
@@ -0,0 +1,28 @@
...
+        Pattern pattern = Pattern.compile(PATH_PARAM_PATTERN);
+
+        Set<String> pathParams = stream(pathSegments)
+                .filter(pathSegment -> pathSegment.matches(PATH_PARAM_PATTERN))
+                .map(pathSegment -> getPathVariable(pattern.matcher(pathSegment)))
+                .collect(Collectors.toSet());
+        return pathParams;
+    }
+
+    private String getPathVariable(Matcher matcher) {
```
Comment (matt-rich) [inline clients/rest-client-core/src/main/java/uk/gov/justice/services/clients/core/RestClientHelper.java]: final method param

File: clients/rest-client-core/src/main/java/uk/gov/justice/services/clients/core/QueryParam.java [hunk 25]
```diff
@@ -0,0 +1,20 @@
+package uk.gov.justice.services.clients.core;
+
+public class QueryParam {
+
+    private String name;
+    private boolean required;
+
+    public QueryParam(String name, boolean required) {
```
Comment (matt-rich) [inline clients/rest-client-core/src/main/java/uk/gov/justice/services/clients/core/QueryParam.java]: final method param

File: clients/rest-client-core/src/main/java/uk/gov/justice/services/clients/core/RestClientHelper.java [hunk 26]
```diff
@@ -0,0 +1,28 @@
...
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import static java.util.Arrays.stream;
+
+public class RestClientHelper {
+    private static final String PATH_PARAM_PATTERN = "\\{(\\w+)\\}";
+
+    public Set<String> extractPathParametersFromPath(String path) {
+        String[] pathSegments = path.split("/");
```
Comment (mapingo) [inline clients/rest-client-core/src/main/java/uk/gov/justice/services/clients/core/RestClientHelper.java]: final in both arguments and variables

File: clients/rest-client-core/src/main/java/uk/gov/justice/services/clients/core/QueryParam.java [hunk 27]
```diff
@@ -0,0 +1,20 @@
+package uk.gov.justice.services.clients.core;
+
+public class QueryParam {
+
+    private String name;
```
Comment (matt-rich) [inline clients/rest-client-core/src/main/java/uk/gov/justice/services/clients/core/QueryParam.java]: final variables, shouldn't change once defined by constructor.

File: clients/rest-client-core/src/main/java/uk/gov/justice/services/clients/core/EndpointDefinition.java [hunk 28]
```diff
@@ -0,0 +1,35 @@
...
+import java.util.Set;
+
+public class EndpointDefinition {
+
+    private String baseURi;
+    private String path;
+    private Set<String> pathParams;
+    private Set<QueryParam> queryParams;
+
+    public EndpointDefinition(String baseURi, String path, Set<String> pathParams,
```
Comment (matt-rich) [inline clients/rest-client-core/src/main/java/uk/gov/justice/services/clients/core/EndpointDefinition.java]: final method param

File: core/src/test/java/uk/gov/justice/services/core/annotation/ComponentTest.java (hunk 3 above)
Comment (ZeeshanGhalib) [inline core/src/test/java/uk/gov/justice/services/core/annotation/ComponentTest.java]: Done

File: clients/rest-client-generator/src/test/java/uk/gov/justice/services/clients/rest/generator/RestClientGenerator_CodeStructureTest.java (hunk 9 above)
Comment (purple52) [inline clients/rest-client-generator/src/test/java/uk/gov/justice/services/clients/rest/generator/RestClientGenerator_CodeStructureTest.java]: It is often subtly different for each test, so I think its best to keep it inline so it is clear what each test is doing

File: clients/rest-client-generator/src/test/java/uk/gov/justice/services/clients/rest/generator/RestClientGenerator_MethodBodyTest.java (hunk 8 above)
Comment (purple52) [inline clients/rest-client-generator/src/test/java/uk/gov/justice/services/clients/rest/generator/RestClientGenerator_MethodBodyTest.java]: Possibly, but as above the tests tend to be subtly different so hard to achieve without obscuring what is being tested

**Pull request merged**